### PR TITLE
Feat: Configuration for mapping saml attributes to user properties

### PR DESCRIPTION
**Pull request by Millefeuille42**
https://github.com/tracim/tracim/issues/6209
https://github.com/tracim/tracim/issues/6210
https://github.com/tracim/tracim/issues/6212

This PR comes from https://github.com/tracim/tracim/pull/6232 so they need to be merged in order, ideally

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [ ] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [x] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [ ] The code is clear enough
- [ ] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [ ] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [ ] Manual, quality tests have been done

### CODE REVIEW

**Comment by Millefeuille42** [comment]
@lebouquetin Considering that this PR is from #6232, should the merge target be `feat/6231__implement_saml_authentication_flow` or `develop` ?

In the latter, this means #6232 needs to be merged in `develop` beforehand despite missing crucial elements implemented in this PR

**Review by lebouquetin — COMMENTED** [review]
@Millefeuille42 the source code looks like a forked version of tracim not intended to be merge in main branch. Am I wrong?

For example, the diff includes this:

```
- className='loginpage__main__form__forgot_password'
- href={'/saml/sso?target=' + encodeURIComponent(idp)}
+ className='loginpage__main__form__saml'
+ href={'/saml/sso?target=' + encodeURIComponent(idp.identifier)}
```

File: backend/settings_saml2.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
                 "want_name_id": true
             }
         },
+        "allow_unknown_attributes": true,
```
[the remark this answers is not in the document]
Comment (lebouquetin) [reply backend/settings_saml2.json:44]: @Millefeuille42 I asked another way: is there a reason we decide to refuse unknown attributes?

**Comment by Millefeuille42** [comment]
@lebouquetin this kind of diffs are here because this PR merges into [feat/6231__implement_saml_authentication_flow](https://github.com/tracim/tracim/tree/feat/6231__implement_saml_authentication_flow) instead of develop.

This is because it is not "enough" as it is to be merged into main/develop. It is an incomplete implementation of SAML and requires further development provided in other branches.

File: backend/settings_saml2.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Millefeuille42) [reply backend/settings_saml2.json:44]: I've not found any reason to refuse unknown attributes to be fair. I do think that this option is here to prevent the parsing + loading in memory of unused attributes (for optimization)

[the remark this answers is not in the document]
Comment (lebouquetin) [reply backend/settings_saml2.json:44]: ok.

File: backend/tracim_backend/config.py [hunk 2]
```diff
@@ -156,6 +156,20 @@ def _get_protected_value(self, value: str, secret: bool) -> str:
             return value
 
 
+class IdPConfig(object):
+    def __init__(self, displayed_name: str, identifier: str, logo_url: str):
```
Comment (lebouquetin) [inline backend/tracim_backend/config.py:160]: need a description explaining that this is related to SAML / Shibbolet identification/authentication

File: backend/tracim_backend/config.py [hunk 3]
```diff
@@ -173,7 +187,7 @@ def __init__(self, settings: typing.Dict[str, typing.Any]):
         self._check_consistency()
         self.check_config_validity()
         logger.debug(self, "CONFIG_PROCESS:5: End of config process")
-        self.IDP_LIST: typing.List[str] = []
+        self.IDP_LIST: typing.List[IdPConfig] = []
```
Comment (lebouquetin) [inline backend/tracim_backend/config.py:190]: I suggest to name this with a more explicit name; something like SAML_IDP_LIST? (is this related to SAML or to SHIBB?)

File: backend/tracim_backend/lib/core/system.py [hunk 4]
```diff
@@ -85,7 +88,7 @@ def get_config(self) -> ConfigModel:
             call__enabled=self._config.CALL__ENABLED,
             call__unanswered_timeout=self._config.CALL__UNANSWERED_TIMEOUT,
             auth_types=auth_types,
-            idp_list=self._config.IDP_LIST,
+            idp_list=idp_list,
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/core/system.py:91]: why not a lambda for a more explicit/direct code read?

```
idp_list=[idp.to_dict() for idp in self._config.IDP_LIST]
```

What is the need for an intermediate variable? debugging?

File: backend/tracim_backend/lib/utils/authentication.py [hunk 5]
```diff
@@ -56,23 +58,38 @@ def __init__(self, app_config: CFG, configurator: Configurator) -> None:
...
-            self.saml_config.attribute_map = attribute_map
         self._metadata_response = Response(content_type="application/xml")
         self._metadata_response.text = create_metadata_string(
             config=self.saml_config, configfile=None
         ).decode()
         self.saml_client = Saml2Client(config=self.saml_config)
 
-        for _, data in self.saml_client.config.vorg.items():
+        self.saml_idp_config = {
+            "default": {
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/utils/authentication.py:68]: move this dict to a global static var on top of authentication.py file: `SAML_IDP_DEFAULT_CONFIG = { ... }`

File: backend/tracim_backend/lib/utils/authentication.py [hunk 6]
```diff
@@ -168,6 +188,8 @@ def _idp_chooser(self, request: TracimRequest) -> Response:
         )
 
     def _acs(self, request: TracimRequest) -> Response:
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/utils/authentication.py:190]: what is acs? I'd like some explanation of what it implements (the name is not self-explicit). Maybe link to related specification

File: backend/tracim_backend/lib/utils/authentication.py [hunk 7]
```diff
@@ -177,15 +199,29 @@ def _acs(self, request: TracimRequest) -> Response:
...
             return Response(e.__str__())
         identity = authn_response.get_identity()
-        if identity is None or "UserID" not in identity:
+        if identity is None:
             return HTTPBadRequest()
+        idp_name = request.POST.get("RelayState")
+
+        formatted_attributes = {}
+        for key, value in self.saml_idp_config[idp_name]["attribute_map"].items():
+            placeholders = re.findall(r"\${(.*?)}", value)
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/utils/authentication.py:208]: /!\ regex: you must explain what pattern you expect to match. Give some examples working (and some not working if there are)

File: backend/tracim_backend/lib/utils/authentication.py [hunk 8]
```diff
@@ -177,15 +199,29 @@ def _acs(self, request: TracimRequest) -> Response:
         except ResponseLifetimeExceed as e:
             return Response(e.__str__())
         identity = authn_response.get_identity()
-        if identity is None or "UserID" not in identity:
+        if identity is None:
             return HTTPBadRequest()
+        idp_name = request.POST.get("RelayState")
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/utils/authentication.py:204]: what is `RelayState`?

File: backend/tracim_backend/lib/utils/authentication.py (hunk 7 above)
Comment (lebouquetin) [reply backend/tracim_backend/lib/utils/authentication.py:208]: note: probably the right way to explain the regex is a comment before the for loop explaining that you are parsing configuration in order to match properties with complex ones. Give somes examples so the developer reading your code immediately understand what is the goal.

File: backend/tracim_backend/lib/utils/authentication.py (hunk 8 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/utils/authentication.py:204]: RelayState is information conveyed through the auth process, this is here to keep info about at which idp the user authenticated. To then get the proper attribute bindings. This usage is standard to SAML

File: backend/tracim_backend/models/context_models.py [hunk 9]
```diff
@@ -85,7 +85,7 @@ def __init__(
         call__enabled: bool,
         call__unanswered_timeout: int,
         auth_types: List[str],
-        idp_list: List[str],
+        idp_list: List[Dict[str, str]],
```
Comment (lebouquetin) [inline backend/tracim_backend/models/context_models.py:88]: rename `idp_list` to `saml_idp_list` or `shibbolet_idp_list`

File: backend/tracim_backend/lib/utils/authentication.py (hunk 6 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/utils/authentication.py:190]: ACS is a term specific to SAML, however I could indeed add some comments above to explain

File: backend/tracim_backend/views/core_api/schemas.py [hunk 10]
```diff
@@ -2139,6 +2139,12 @@ class RoleSchema(marshmallow.Schema):
     label = marshmallow.fields.String(required=True, example="Reader")
 
 
+class IdPConfigSchema(marshmallow.Schema):
```
Comment (lebouquetin) [inline backend/tracim_backend/views/core_api/schemas.py:2142]: Rename `IdPConfigSchema` to `SAMLIdPConfigSchema` or `ShibboletIdpConfigSchema` or something clear that it is important for SAML/Shibb authentication only (and you dont care in "classical" case)

File: backend/tracim_backend/views/core_api/schemas.py [hunk 11]
```diff
@@ -2168,7 +2174,7 @@ class ConfigSchema(marshmallow.Schema):
     call__enabled = marshmallow.fields.Bool()
     call__unanswered_timeout = marshmallow.fields.Int()
     auth_types = marshmallow.fields.List(marshmallow.fields.String())
-    idp_list = marshmallow.fields.List(marshmallow.fields.String())
+    idp_list = marshmallow.fields.Nested(IdPConfigSchema, many=True)
```
Comment (lebouquetin) [inline backend/tracim_backend/views/core_api/schemas.py:2177]: same: rename `idp_list`

File: backend/tracim_backend/lib/core/system.py (hunk 4 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/core/system.py:91]: IIRC this didn't work for some reason, will try again as you proposed to be sure

File: doc/backend/setting.md [hunk 12]
```diff
@@ -85,6 +86,132 @@ ldap_tls = False
 ⚠ When logging in Tracim, if a valid LDAP user doesn't
 exist in Tracim, it will be created as a standard user.
 
+## SAML Authentication
```
Comment (lebouquetin) [inline doc/backend/setting.md:89]: I suggest to move this to end of authentication related stuff.
API-Key will be a more common use case than SAML, so SAML can be moved at the end of section.

File: frontend/src/component/Login/SignIn.jsx [hunk 13]
```diff
@@ -38,13 +38,20 @@ const SignIn = props => {
             </Link>
           </li>
           {props.system.config.idp_list.map((idp) =>
-            <li key={idp}>
+            <li key={idp.identifier}>
```
Comment (lebouquetin) [inline frontend/src/component/Login/SignIn.jsx:41]: idp is a too generic naming, I think.

File: frontend/src/component/Login/SignIn.jsx [hunk 14]
```diff
@@ -38,13 +38,20 @@ const SignIn = props => {
...
               <a
-                className='loginpage__main__form__forgot_password'
-                href={'/saml/sso?target=' + encodeURIComponent(idp)}
+                className='loginpage__main__form__saml'
```
Comment (lebouquetin) [inline frontend/src/component/Login/SignIn.jsx:43]: I do not understand why you replace `forgot password` with `saml` in a generic component. Do I understand wrong?

Comment (lebouquetin) [reply frontend/src/component/Login/SignIn.jsx:43]: After investigation, I feel like the diff is not the real diff (you diff with some already included SAML related code, isn't it?)

Comment (Millefeuille42) [reply frontend/src/component/Login/SignIn.jsx:43]: the `forgot_password` class was used because it initially looked OK for this usage. In this PR since I changed the look of the IdP list I created specific classes for SAML IdP list

Comment (Millefeuille42) [reply frontend/src/component/Login/SignIn.jsx:43]: As said earlier, the diff is with #6232, since it is the current target. So the diff that are shown are related with already included SAML code

File: backend/tracim_backend/lib/utils/authentication.py (hunk 6 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/utils/authentication.py:192]: fixed

File: backend/tracim_backend/lib/utils/authentication.py (hunk 8 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/utils/authentication.py:208]: added details in comments

File: backend/tracim_backend/lib/utils/authentication.py (hunk 7 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/utils/authentication.py:218]: added details in comment

File: backend/tracim_backend/views/core_api/schemas.py (hunk 10 above)
Comment (Millefeuille42) [reply backend/tracim_backend/views/core_api/schemas.py]: done

File: backend/tracim_backend/views/core_api/schemas.py (hunk 11 above)
Comment (Millefeuille42) [reply backend/tracim_backend/views/core_api/schemas.py]: done

File: frontend/src/component/Login/SignIn.jsx (hunk 13 above)
Comment (Millefeuille42) [reply frontend/src/component/Login/SignIn.jsx]: fixed

File: backend/tracim_backend/lib/utils/authentication.py (hunk 5 above)
Comment (Millefeuille42) [reply backend/tracim_backend/lib/utils/authentication.py]: done

File: backend/tracim_backend/config.py (hunk 3 above)
Comment (Millefeuille42) [reply backend/tracim_backend/config.py]: done

File: backend/tracim_backend/config.py (hunk 2 above)
Comment (Millefeuille42) [reply backend/tracim_backend/config.py:160]: renamed it to SamLIdPConfig

Comment (lebouquetin) [reply backend/tracim_backend/config.py:160]: Why `SamLIdPConfig` and not `SAMLIdPConfig` (or `SamlIdpConfig`)?

**Comment by lebouquetin** [comment]
I have a last question about the `SamLIdPConfig` naming, all other stuff is ok for me

File: backend/tracim_backend/config.py (hunk 2 above)
Comment (Millefeuille42) [reply backend/tracim_backend/config.py:160]: For readability.

SAMLIdPConfig is kind of hard to read and SamlIdPConfig is even worse on most fonts since lowercase l and uppercase I can look very similar.

No issue in changing it to a more conventional naming, but I'd prefer to keep it that way so it's not hard to read (at least for me)

Comment (lebouquetin) [reply backend/tracim_backend/config.py:160]: So, I suggest to put an inline comment: `# INFO - M.L - 2023-10-25: the naming SamLIdPConfig has been preferred to SAMLIdpConfig / SamlIdPConfig for readability`

This way, developers will not decide to change this naming in the future

Comment (lebouquetin) [reply backend/tracim_backend/config.py:160]: (and they will understand why)

Comment (Millefeuille42) [reply backend/tracim_backend/config.py:160]: Done and noted !
